### PR TITLE
Asarje/shwetank upsample

### PR DESCRIPTION
**Pull request by mywoodstock**

### CODE REVIEW

File: tt_eager/tt_dnn/op_library/upsample/kernels/dataflow/writer_upsample_multi_core_sharded.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdint.h>
+#include "dataflow_api.h"
+
+#include "debug/dprint.h"
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/upsample/kernels/dataflow/writer_upsample_multi_core_sharded.cpp:8]: Remove?

**Review by TT-BrianLiu — APPROVED** [review]

File: ttnn/ttnn/operations/others.py [hunk 2]
```diff
@@ -438,4 +438,58 @@ def mean(input_tensor: ttnn.Tensor, dim: Union[int, Tuple[int]], keepdim: bool =
...
+
+
+@ttnn.register_operation(
+    name="ttnn.upsample",
+    validate_input_tensors=_upsample_validate_input_tensors,
+    torch_function=_torch_upsample,
+)
+def upsample(
+    input_tensor: ttnn.Tensor,
+    scale_factor: Union[float, Tuple[float, float], Tuple[float, float, float, float]],
```
Comment (arakhmati) [inline ttnn/ttnn/operations/others.py:470]: Why is `scale_factor` a float? It's `int` in `torch`. Do we ever pass in actual floating point values? 

I think the type should be `Union[int, Tuple[int, ...]]`

File: ttnn/ttnn/operations/others.py [hunk 3]
```diff
@@ -438,4 +438,58 @@ def mean(input_tensor: ttnn.Tensor, dim: Union[int, Tuple[int]], keepdim: bool =
...
+    memory_config: ttnn.MemoryConfig = ttnn.DRAM_MEMORY_CONFIG,
+) -> ttnn.Tensor:
+    """
+    upsample(input_tensor: ttnn.Tensor, scale_factor: Union[float, Tuple[float, float], Tuple[float, float, float, float]) -> ttnn.Tensor
+    """
+
+    scale_h, scale_w = 1, 1
+    if isinstance(scale_factor, float):
+        scale_h = 1
```
Comment (arakhmati) [inline ttnn/ttnn/operations/others.py:479]: Shouldn't `scale_h `be `scale_factor` in this case?

File: ttnn/ttnn/operations/others.py [hunk 4]
```diff
@@ -438,4 +438,58 @@ def mean(input_tensor: ttnn.Tensor, dim: Union[int, Tuple[int]], keepdim: bool =
...
+        scale_w = scale_factor
+    elif len(scale_factor) == 2:
+        scale_h, scale_w = scale_factor
+    elif len(scale_factor) == 4:
+        scale_h, scale_w = scale_factor[1], scale_factor[2]
```
Comment (arakhmati) [inline ttnn/ttnn/operations/others.py:484]: There needs to be a check to make sure that `scale_factor[0]` and `scale_factor[3]` is equal to `1`

File: ttnn/ttnn/operations/others.py [hunk 5]
```diff
@@ -438,4 +438,58 @@ def mean(input_tensor: ttnn.Tensor, dim: Union[int, Tuple[int]], keepdim: bool =
...
+        input_tensor,
+        ranks=(2, 3, 4),
+        dtypes=(ttnn.bfloat16, ttnn.bfloat8_b),
+        layouts=(ttnn.ROW_MAJOR_LAYOUT,),
+        can_be_on_device=True,
+        can_be_on_cpu=False,
+    )
+
+
+def _torch_upsample(input_tensor: ttnn.Tensor, scale_factor: [float, float], **_):
```
Comment (arakhmati) [inline ttnn/ttnn/operations/others.py:453]: The type of `scale_factor` here should match the one in `def upsample`

**Review by arakhmati — CHANGES_REQUESTED** [review]

File: ttnn/ttnn/operations/others.py (hunk 3 above)
Comment (mywoodstock) [reply ttnn/ttnn/operations/others.py]: yes, updated.

**Review by arakhmati — APPROVED** [review]

**Pull request merged**